### DISCUSSION TITLE
useQuery with possibly undefined inputs

### DISCUSSION THREAD

**Discussion by WesleyYue** [trpc/trpc#3529] [Q&A] [answered]
Let's say I want to use a trpc useQuery with a React state variable that is possibly undefined. Since trpc enforces type safety, it'll give me an error because my possibly undefined variable does not match with the trpc expected input type. Like this:

```
function Component(): JSX.Element {
  const { ourUser } = useContext(UserContext);  // `ourUser` type is `string | null`
  const userQuery = trpc.user.useQuery(
    {
      user: ourUser.id,  // Error here: `user` type is `string`, whereas ourUser.id is possibly null or undefined.
    },
    {
      enabled: !isNullOrUndefined(ourUser?.id),
    }
  );
  return (<>...</>);
}
```

This ts error occurs because ts doesn't know that the assignment is conditional behind the `enabled` option. So I end up changing it to something like `user: ourUser?.id ?? "",` to satisfy the compiler, even though user will never be "". This makes the code a bit less intuitive to read. Is there a more ergonomic way to achieve this without having a conditional on the input line with a fake placeholder input ("")?

**Comment by iway1** [comment, marked as answer]
I don't think there is an easier way in general... This is kind of quirk with react query in general any time it's used in a typesafe way. 

It'd be nice if the typescript types were such that id can be defined if `enabled: false`

**Reply by KATT** [reply]
Yeah, you gotta typecast the input here, no way around it until we can have conditional hooks
